toolchain/check/cpp/type_mapping.cpp

```diff
@@ -180,6 +180,36 @@ static auto MapNonWrapperType(Context& context, SemIR::InstId inst_id,
   }
 }
 
+// Returns `void*` if the type is a wrapped `Cpp.void*`, consuming the pointer
+// from `wrapper_types`. Otherwise returns no type.
+static auto TryMapVoidPointer(Context& context, SemIR::TypeId type_id,
+                              llvm::SmallVector<SemIR::TypeId>& wrapper_types)
+    -> clang::QualType {
+  if (type_id != SemIR::CppVoidType::TypeId) {
+    return clang::QualType();
+  }
+
+  llvm::SmallVector<SemIR::TypeId>::iterator pointer_iter;
+  if (!wrapper_types.empty() &&
+      context.types().Is<SemIR::PointerType>(wrapper_types.back())) {
+    // `void*`.
+    pointer_iter = wrapper_types.end() - 1;
+  } else if (wrapper_types.size() >= 2 &&
+             context.types().Is<SemIR::ConstType>(wrapper_types.back()) &&
+             context.types().Is<SemIR::PointerType>(
+                 wrapper_types[wrapper_types.size() - 2])) {
+    // `const void*`.
+    pointer_iter = wrapper_types.end() - 2;
+  } else {
+    return clang::QualType();
+  }
+
+  wrapper_types.erase(pointer_iter);
```

Comment:
An alternative way to write this, which I'm not sure is better:

```suggestion
  llvm::SmallVector<SemIR::TypeId>::iterator wrapper_iter =
      wrapper_types.end() - 1;
  if (context.types().Is<SemIR::PointerType>(*wrapper_iter)) {
    // This is `void*`, wrapper_iter points to the pointer type.
  } else if (wrapper_iter != wrapper_types.begin() &&
             context.types().Is<SemIR::ConstType>(*wrapper_iter) &&
             context.types().Is<SemIR::PointerType>(*(wrapper_iter - 1))) {
    // This is `const void*`, move `wrapper_iter` to point to the pointer type.
    --wrapper_iter;
  } else {
    return clang::QualType();
  }

  wrapper_types.erase(wrapper_iter);
```